Refactor capture.py to be a class

d3ming opened the pull request:
1. Make it more testable now that it can do phantom and remote webdrivers
2. (more importantly) hook up tagtester to be able to use phantomjs

d3ming committed:
Refactor capture.py to be a class

1. Make it more testable now that it can do phantom and remote webdrivers
2. (more importantly) hook up tagtester to be able to use phantomjs

bitglue commented on tagcompare/capture.py:
```diff
@@ -14,11 +17,164 @@
+                self._driver, tag_element, output_path)
+        except selenium.common.exceptions.WebDriverException:
+            LOGGER.exception("Exception while capturing tags!")
+            return False
```
Returning `False` on errors is not very pythonic. Why not just raise an exception? Either let this one be raised and log it on the way out, just let in be raised and don't even try to handle it here, or catch this exception and raise another one.

d3ming replied:
Will fix

bitglue commented on tagcompare/capture.py:
```diff
@@ -14,11 +17,164 @@
+                        if r is None:
+                            num_existing_skipped += 1
+                        elif r is False:
+                            continue
```
What's the difference between `None` and `False`?

I think I know the answer: `False` means there was an error. But that will not be obvious to future readers seeing this out of context, nor is it something typically expected in Python.

bitglue replied:
Future readers that guess `False` is an error will also wonder why you are doing nothing about it. Perhaps you do just want to log it, but if you do that here, instead of in `_capture_tag`, then the `log(...); continue` says very clearly "I intend to log this error and then continue". The intent is much less clear, split up as it is now.

d3ming replied:
The problem is that I want to log an aggregated number of errors/skips, which I currently do in the `capture_tags` method:

```
        LOGGER.info(
            "Captured %s tags, skipped %s existing tags for config=%s.  Found %s errors",
            num_captured, num_existing_skipped, self._caps, len(browser_errors))
```

Is your suggestion to return `None` on skips and just let it throw on error and have `capture_tags` catch the exception and increment the error count?

bitglue replied:
I was thinking you'd not worry about handling the exception in `_capture_tag()`, because it doesn't really know how to handle them. Let them propagate here, and handle them here, because here you know what to do with them:
- log the failure
- increment the error counter
- move on to the next thing

d3ming replied:
Got it

bitglue commented on tagcompare/capture.py:
```diff
@@ -14,11 +17,164 @@
+            self._driver = None
+        finally:
+            if self._driver:
+                self._driver.quit()
```
The `finally:` and `except KeyboardInterrupt:` block here seem very similar. Why are they different?

d3ming replied:
Yeah good point, I'll consolidate

bitglue commented on tagcompare/capture.py:
```diff
@@ -14,11 +17,164 @@
+        self._driver = driver
+        self._caps = caps
+
+    def stop(self):
```
You might consider calling this `close()`, which is something of a convention and would allow `TagCompare` instances to be used with `contextlib.closing`, for example.

d3ming replied:
Will do

bitglue commented on tagcompare/capture.py:
```diff
@@ -175,7 +225,8 @@ def main():
     configs = settings.DEFAULT.configs_in_comparisons()
-    __capture_tags_for_configs(cids=cids, pathbuilder=pathbuilder, configs=configs)
+    _capture_tags_for_configs(
+        cids=cids, pathbuilder=pathbuilder, configs=configs)
```
Does `return original_build` have any effect here?

Also, if `_capture_tags_for_configs` encounters errors, should the program exit with non-zero exit status? It looks like no, although I'm guessing it should. I would recommend again raising exceptions instead of returning values that indicate an error: it will make this sort of thing much harder to miss.

d3ming replied:
The original build is used when we chain capture.py with compare.py, which used to be the main scenario for tagcompare.  In this case they need to share the same build to do compare against and we aren't returning this as a means to indicate error or not (see https://github.com/paperg/tagcompare/blob/master/tagcompare/main.py#L110-L112).

bitglue replied:
I see. What can we do to make that more clear? Can the function be named something to suggest that a build is returned? Or can it be written in the docstring?

Or perhaps better: should the build be an object, and this should be a method on it?

d3ming replied:
Good suggestion, I'll add a docstring for now.  Not sure if it's worth it at this point to change beyond that (certainly outside the scope of this PR)

bitglue commented on tagcompare/compare.json:
```diff
@@ -16,6 +16,9 @@
     },
+    "phantomjs": {
+      "enabled": true
+    },
```
Under what circumstances would this configuration be altered? Is this [soft coding](http://thedailywtf.com/articles/Soft_Coding)?

d3ming replied:
Are you talking about the `enabled` part or this whole file in general?

bitglue replied:
I was thinking mostly the whole file. Reading it, I'm not entirely sure what it does, or why it's needed. Usually when I see config files, I think about why I'd want to edit them. Like, I want to edit /etc/apache.conf because I don't want to recompile Apache to change my website configuration.

Maybe this file has a good use, but I can't tell and that's also a problem. If you want to keep it, what do you think about using YAML so you can at least have comments?

d3ming replied:
Yeah it's not exactly clear because `compare.json` really does 2 things: 1. list all the set of browser configs and 2. list the comparisons you'd wanna do against the configs

(1) doesn't really need to be configurable, is your suggestion to make it part of the code, in say `browserconfigs.py`?
(2) does need to be configurable because the runner of tagcompare should be able to specify locally what kind of comparisons he cares about

Out of scope for this PR though - opened https://github.com/paperg/tagcompare/issues/54

bitglue commented on tagcompare/logger.py:
```diff
@@ -2,12 +2,17 @@
 
 
+def generate_timestamp():
+    return str.format(time.strftime("%Y%m%d-%H%M%S"))
```
Why is `str.format()` needed here?

d3ming replied:
I like to format strings in my spare time for fun, jk - removed

bitglue commented on tagtester/main.py:
```diff
@@ -61,10 +58,11 @@ def capture_tags():
                         output_path=pb.tagimage, tagtype=t)
-        driver.quit()
+                    LOGGER.info('Captured {}'.format(pb.tagimage))
+        tagcapture.stop()
```
You probably want this in a `finally:` block so it happens if there's an exception. Or rename it to `close()` and use `contextlib.closing`, like:

``` python
from contextlib import closing
with closing(capture.TagCapture.from_config(config)) as tagcapture:
    # do your stuff with tagcapture, confident that it will be closed
    # when you are done, regardless of what happens.
```

d3ming replied:
Oh I like this contextlib thing

d3ming commented:
@bitglue addressed most of the comments with https://github.com/paperg/tagcompare/pull/53/commits/4f571523110dc8ed152c781d15348a596a357a6f

Code coverage is now 94%!  (up from 93)

```
---------------------------------- coverage: platform darwin, python 2.7.8-final-0 -----------------------------------
Name                       Stmts   Miss  Cover   Missing
--------------------------------------------------------
tagcompare/__init__.py         0      0   100%
tagcompare/capture.py        132     21    84%   94-95, 123-125, 133-135, 139, 141, 168, 185-186, 223-235
tagcompare/compare.py        126     23    82%   33, 44-45, 61-62, 113-114, 140-141, 143-144, 186-188, 195-206
tagcompare/image.py           88      4    95%   20, 67, 83, 95
tagcompare/logger.py          46      0   100%
tagcompare/output.py         173      6    97%   253-262, 333
tagcompare/placelocal.py      94      0   100%
tagcompare/settings.py       138      0   100%
tagcompare/webdriver.py       94      0   100%
--------------------------------------------------------
TOTAL                        891     54    94%
```
